sycl/source/detail/event_impl.hpp

```diff
@@ -200,11 +205,16 @@ class event_impl {
   bool MHostEvent = true;
   std::unique_ptr<HostProfilingInfo> MHostProfilingInfo;
   void *MCommand = nullptr;
+  std::weak_ptr<queue_impl> MQueue;
 
   /// Dependency events prepared for waiting by backend.
   std::vector<EventImplPtr> MPreparedDepsEvents;
   std::vector<EventImplPtr> MPreparedHostDepsEvents;
 
+  /// Indicates that the task associated with this event has been submitted by
+  /// the queue to the device.
+  std::atomic<bool> MIsFlushed = false;
```

Comment:
`flushIfNeeded()` isn't executed atomically. Hence, I believe, there's a need for a more sophisticated synchronization than a mere atomic.
Otherwise, if call to `flushIfNeeded()` is guarded by a lock it should be feasible to let this flag be a mere bool rather than atomic.

Reply:
Are there any sort of commands that make the queue flushed implicitly? Is there any handling of such commands so that `MIsFlushed` becomes true during init?

Reply:
Discussed both of these separately. To summarize:
1. There are no functional issues with this approach since the used PI calls are thread-safe. And from performance standpoint, worst case scenario is that we'll call `piQueueFlush` twice.
2. There are, `piQueueRelease`, `piEventsWait` and blocking enqueue commands. `piQueueRelease` is detected by a nullptr QueueImplPtr, while the other cases are not handled to avoid execution graph traversal during enqueueing.